File: clients/kotlin/src/main/kotlin/com/priostack/acn/examples/Quickstart.kt
/*
 * Quickstart: register an agent, store facts, and read them back.
 *
 * It talks to the live ACN and registers a throwaway agent. Run it with:
 *
 *     ./gradlew run
 *     ./gradlew run --args="my-agent-name"
 *
 * The network calls are guarded behind `main`, so depending on the library
 * (`com.priostack:priostack-acn`) never fires a request.
 */
package com.priostack.acn.examples

import com.priostack.acn.AcnClient
import com.priostack.acn.ObjectType
import com.priostack.acn.StoreObject

fun main(args: Array<String>) {
    val displayName = args.firstOrNull() ?: "kotlin-quickstart"

    // `use` disconnects the session and releases resources on exit.
    AcnClient().use { acn ->
        // 1. Self-register (no API key, no dashboard). The token is captured on
        //    the client and returned so you can persist it for next time.
        val reg = acn.register(displayName)
        println("1. registered agent ${reg.agentId} (tier ${reg.tier})")
        println("   token (store this, shown once): ${reg.token}")

        // 2. Open a session. The session id is captured internally.
        val conn = acn.connect()
        println("2. session ${conn.sessionId} | capabilities ${conn.grantedCapabilities}")

        // 3. Create an isolated memory space. Use the RETURNED id for writes.
        val space = acn.createSpace("prod-agent-memory")
        println("3. created space ${space.spaceId}")

        // 4. Persist typed facts (declaration | observation | measurement).
        val stored = acn.store(
            space.spaceId,
            listOf(
                StoreObject("User approved execution workflow #402.", ObjectType.DECLARATION),
                StoreObject("Export pipeline latency was 1.8s at 14:02 UTC.", ObjectType.OBSERVATION),
            ),
        )
        println("4. stored ${stored.count} objects")

        // 5. Read them back. `query` is a case-insensitive substring filter.
        val hits = acn.fetch(space.spaceId, query = "workflow")
        println("5. fetched ${hits.matched}/${hits.total} (viewport ${hits.returnedTokens} tokens):")
        hits.contents().forEach { println("     - $it") }
    }
}

File: clients/scala/src/main/scala/com/priostack/acn/examples/Quickstart.scala
package com.priostack.acn.examples

import com.priostack.acn.*

/** End-to-end quickstart: register -> connect -> createSpace -> store -> fetch.
  *
  * All live calls happen inside `main`, so importing the library never fires a
  * network request. Run it against the live endpoint with:
  *
  * {{{
  *   sbt "runMain com.priostack.acn.examples.Quickstart scala-sdk-smoke"
  * }}}
  *
  * The first CLI arg overrides the display name (default `scala-quickstart`);
  * set `PRIOSTACK_ENDPOINT` to point at a different ACN.
  */
object Quickstart:
  def main(args: Array[String]): Unit =
    val displayName = args.headOption.getOrElse("scala-quickstart")
    val endpoint    = sys.env.getOrElse("PRIOSTACK_ENDPOINT", AcnClient.DefaultEndpoint)

    val acn = AcnClient(endpoint = endpoint)
    try
      val reg = acn.register(displayName)
      println(s"registered: agent=${reg.agentId} account=${reg.accountId} tier=${reg.tier}")

      val conn = acn.connect()
      println(s"connected:  session=${conn.sessionId} account=${conn.resolvedAccount}")
      println(s"            capabilities=${conn.grantedCapabilities.mkString("[", ", ", "]")}")

      val space = acn.createSpace("scala-quickstart-space")
      println(s"created:    space=${space.spaceId}")

      val stored = acn.store(
        space.spaceId,
        Seq(
          StoreObject.declaration("Refunds over $500 need manager approval."),
          StoreObject.observation("Customer #402 triggered the export pipeline.")
        )
      )
      println(s"stored:     ${stored.count} object(s)")

      val hits = acn.fetch(space.spaceId, query = "refund")
      println(s"fetched:    matched=${hits.matched} total=${hits.total} returnedTokens=${hits.returnedTokens}")
      hits.contents.foreach(c => println(s"  - $c"))
    catch
      case e: AcnToolError =>
        Console.err.println(s"[tool error: ${e.outcome}] ${e.detail}")
        sys.exit(1)
      case e: AcnTransportError =>
        Console.err.println(s"[transport error] ${e.getMessage}")
        sys.exit(1)
    finally acn.close()
